### PR TITLE
Neighbor Count Functionality for Memory/Distributed Graphs

### PR DESCRIPTION
**Pull request by adithyabhonsley**
- [X ] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [X ] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [X ] Documentation is added or updated to reflect new code in the same format as the rest of the repo.
- [ X] PR is labeled using the label menu on the right side.

Previous Behavior
----------------
* Provide relevant issue number if applicable, eg #44.

New Behavior
----------------
* Adds neighbor count (node degree) functionality for both memory graph and distributed graphs
* Adds testing for neighbor count feature for memory/distributed graph (base cases, single/multiple partition, invalid nodes and types)
* Adds Python handling of c functions leveraging ctypes library

### CODE REVIEW

File: src/cc/lib/py_graph.cc [hunk 1]
```diff
@@ -621,6 +621,42 @@ int32_t GetNeighborsInternal(PyGraph *py_graph, NodeID *in_node_ids, size_t in_n
...
 
+int32_t NeighborCount(PyGraph *py_graph, NodeID *in_node_ids, size_t in_node_ids_size, Type *in_edge_types,
+                      size_t in_edge_types_size, uint64_t *out_neighbor_counts)
+{
+    if (py_graph->graph == nullptr)
+    {
+        RAW_LOG_ERROR("Internal graph is not initialized");
+        return 1;
+    }
+    std::fill_n(out_neighbor_counts, in_node_ids_size, 0);
```
Comment (coledie) [inline src/cc/lib/py_graph.cc:632]: Is it possible to avoid this fill?

File: .bazelrc [hunk 2]
```diff
@@ -15,8 +15,7 @@ build:manylinux --crosstool_top=//tools/toolchain:manylinux
 build:manylinux --cpu=k8
 build:manylinux --host_crosstool_top=@bazel_tools//tools/cpp:toolchain
 
-build:darwin --action_env=BAZEL_LINKLIBS=-l%:libstdc++.a
-build:darwin --action_env=BAZEL_LINKOPTS=-static-libgcc:-lm:-pthread
+build:darwin --action_env=BAZEL_LINKOPTS=-lm:-pthreadj
```
Comment (mortid0) [inline .bazelrc:18]: why this change is needed for neighbor count feature?

File: src/cc/lib/graph/partition.cc [hunk 3]
```diff
@@ -441,9 +441,9 @@ void Partition::GetNodeStringFeature(uint64_t internal_node_id, std::span<const
     }
 }
 
-size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_types,
-                               std::vector<NodeId> &out_neighbors_ids, std::vector<Type> &out_edge_types,
-                               std::vector<float> &out_edge_weights) const
+template <class F>
+size_t Partition::FetchNeighborInfo(uint64_t internal_id, std::span<const Type> edge_types, F func) const
```
Comment (coledie) [inline src/cc/lib/graph/partition.cc:445]: Could we rename this to "NeighborIndexIterator" or something more apt, it seems this is being used more than just to get info.

File: src/cc/lib/graph/partition.cc [hunk 4]
```diff
@@ -476,24 +476,44 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
+
+            result += func(start, last, i);
         }
     }
     return result;
 }
+
+size_t Partition::NeighborCount(uint64_t internal_id, std::span<const Type> edge_types) const
+{
+    auto lambda = [&](auto start, auto last, int i) { return last - start; };
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:488]: please, don't capture everything by reference in lambda, but explicitly list items you'll use.

Comment (mortid0) [reply src/cc/lib/graph/partition.cc:488]: Same below

File: src/cc/lib/py_graph.cc [hunk 5]
```diff
@@ -621,6 +621,42 @@ int32_t GetNeighborsInternal(PyGraph *py_graph, NodeID *in_node_ids, size_t in_n
...
+    if (py_graph->graph->graph)
+    {
+        py_graph->graph->graph->NeighborCount(
+            std::span(reinterpret_cast<snark::NodeId *>(in_node_ids), in_node_ids_size),
```
Comment (coledie) [inline src/cc/lib/py_graph.cc:636]: Are all these casts needed?

File: src/python/deepgnn/graph_engine/snark/tests/BUILD [hunk 6]
```diff
@@ -42,8 +42,8 @@ py_test(
         requirement("networkx"),
         requirement("opencensus"),
         requirement("grpcio"),
-        requirement("fsspec"),
         requirement("grpcio-health-checking"),
+        requirement("fsspec"),
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/tests/BUILD:46]: why this change is needed?

**Review by mortid0 — COMMENTED** [review]
Overall looks great, please, remove empty MANIFEST.in file.

**Review by coledie — CHANGES_REQUESTED** [review]
Looks good

File: WORKSPACE [hunk 7]
```diff
@@ -17,16 +17,16 @@ http_archive(
...
+    strip_prefix = "rules_python-0.9.0",
+    url = "https://github.com/bazelbuild/rules_python/archive/refs/tags/0.9.0.tar.gz",
 )
 
 load("@rules_python//python:pip.bzl", "pip_parse")
 
 pip_parse(
     name = "pip_deps",
-    python_interpreter = "python",
+    python_interpreter = "python3",
```
Comment (mortid0) [inline WORKSPACE:29]: Please, keep python interpreter

File: src/cc/lib/py_graph.cc [hunk 8]
```diff
@@ -621,6 +621,41 @@ int32_t GetNeighborsInternal(PyGraph *py_graph, NodeID *in_node_ids, size_t in_n
...
+    {
+        py_graph->graph->client->NeighborCount(
+            std::span(reinterpret_cast<snark::NodeId *>(in_node_ids), in_node_ids_size),
+            std::span(reinterpret_cast<snark::Type *>(in_edge_types), in_edge_types_size),
+            std::span(out_neighbor_counts, in_node_ids_size));
+        return 0;
+    }
+    catch (const std::exception &e)
+    {
+        RAW_LOG_ERROR("Exception while sampling neighbors: %s", e.what());
```
Comment (mortid0) [inline src/cc/lib/py_graph.cc:652]: Fix error message, looks like it was copy-pasted.

File: src/cc/tests/distributed_test.cc [hunk 9]
```diff
@@ -413,6 +413,57 @@ TEST(DistributedTest, UniformSampleNeighborsWithoutReplacementMultipleServers)
...
+    std::vector<snark::NodeId> input_nodes = {0};
+    std::vector<snark::Type> input_types = {0, 1};
+    std::vector<uint64_t> output_counts(input_nodes.size());
+    c.NeighborCount(std::span(input_nodes), std::span(input_types), std::span(output_counts));
+    EXPECT_EQ(output_counts, std::vector<uint64_t>({3}));
+}
+
+TEST(DistributedTest, NeighborCountMultipleTypesMultipleServers)
+{
+    auto environment = CreateMultiServerEnvironment("FullNeighborsMultipleServers");
```
Comment (mortid0) [inline src/cc/tests/distributed_test.cc:457]: Please, use test name for the environment name.

File: src/cc/tests/distributed_test.cc [hunk 10]
```diff
@@ -413,6 +413,57 @@ TEST(DistributedTest, UniformSampleNeighborsWithoutReplacementMultipleServers)
...
+    auto &c = *environment.second;
+
+    std::vector<snark::NodeId> input_nodes = {0, 55};
```
Comment (mortid0) [inline src/cc/tests/distributed_test.cc:460]: could you add one more node with id=1000 and verify that it has 0 neighbors?

File: src/cc/lib/distributed/client.cc [hunk 11]
```diff
@@ -619,6 +619,56 @@ void GRPCClient::GetEdgeStringFeature(std::span<const NodeId> edge_src_ids, std:
...
+                return;
+            }
+            for (size_t curr_node = 0; curr_node < std::size(output_neighbor_counts); ++curr_node)
+            {
+
+                for (size_t reply_index = 0; reply_index < std::size(replies); ++reply_index)
+                {
+                    const auto &reply = replies[reply_index];
+                    const auto count = reply.neighbor_counts(curr_node);
+                    if (count == 0)
```
Comment (mortid0) [inline src/cc/lib/distributed/client.cc:656]: this if is totally unnecessary.

File: src/cc/lib/graph/partition.cc [hunk 12]
```diff
@@ -441,9 +441,11 @@ void Partition::GetNodeStringFeature(uint64_t internal_node_id, std::span<const
     }
 }
 
-size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_types,
-                               std::vector<NodeId> &out_neighbors_ids, std::vector<Type> &out_edge_types,
-                               std::vector<float> &out_edge_weights) const
+template <class F>
+size_t NeighborIndexIterator(uint64_t internal_id, std::span<const Type> edge_types, F func,
+                             std::vector<uint64_t> m_neighbors_index, std::vector<Type> m_edge_types,
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:446]: Please, don't copy entire graph indices, edge types and type offsets. Pass these vectors by reference instead of value..

**Review by mortid0 — CHANGES_REQUESTED** [review]

File: src/cc/lib/graph/partition.cc [hunk 13]
```diff
@@ -476,24 +478,44 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
-            auto original_type_size = out_edge_types.size();
-            out_edge_types.resize(original_type_size + last - start, m_edge_types[i]);
-            out_edge_weights.reserve(out_edge_weights.size() + last - start);
-            for (size_t index = start; index < last; ++index)
-            {
-                out_edge_weights.emplace_back(index > start ? m_edge_weights[index] - m_edge_weights[index - 1]
-                                                            : m_edge_weights[start]);
-            }
+
+            result += func(start, last, i);
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:482]: It seems like the `result` is always the same `last-start`. So we don't need to wait for the function and update it in place. tl;dr
```c++
result += last-start;
func(start, last, i);
```

File: src/cc/lib/graph/partition.cc [hunk 14]
```diff
@@ -1163,4 +1185,4 @@ Metadata Partition::GetMetadata() const
 {
     return m_metadata;
 }
-} // namespace snark
+} // namespace snark
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:1188]: nit: keep new line?

File: src/cc/tests/graph_test.cc [hunk 15]
```diff
@@ -1004,5 +1004,177 @@ TEST(GraphTest, GetNodeTypesAcrossPartitions)
...
+    // Check for both edge types
+    types = {0, 1};
+    std::fill_n(output_neighbors_count.begin(), 2, 0);
+
+    g.NeighborCount(std::span(nodes), std::span(types), output_neighbors_count);
+    EXPECT_EQ(std::vector<uint64_t>({2, 3}), output_neighbors_count);
+
+    // Check returns 0 for unsatisfying edge types
+    types = {-1, 100};
+    std::fill_n(output_neighbors_count.begin(), 2, 0);
```
Comment (mortid0) [inline src/cc/tests/graph_test.cc:1075]: fill array with -1s to verify it was actually updated, because it can be not touched at all.

**Comment by mortid0** [comment]
@adithyabhonsley Before requesting a review, please rebase on latest main to trigger additional checks and resolve all the comments from previous iterations.
image

File: src/cc/lib/graph/partition.cc [hunk 16]
```diff
@@ -476,24 +478,45 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
-            out_edge_types.resize(original_type_size + last - start, m_edge_types[i]);
-            out_edge_weights.reserve(out_edge_weights.size() + last - start);
-            for (size_t index = start; index < last; ++index)
-            {
-                out_edge_weights.emplace_back(index > start ? m_edge_weights[index] - m_edge_weights[index - 1]
-                                                            : m_edge_weights[start]);
-            }
+
+            result += func(start, last, i);
+            // func(start, last, i);
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:483]: remove commented code.

File: src/cc/lib/graph/partition.cc [hunk 17]
```diff
@@ -476,24 +478,45 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
+            result += last - start;
+            func(start, last, i);
         }
     }
     return result;
 }
+
+size_t Partition::NeighborCount(uint64_t internal_id, std::span<const Type> edge_types) const
+{
+    auto lambda = [](auto start, auto last, int i) { return last - start; };
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:491]: You don't need this result anymore. Pass a function that does nothing.

File: src/cc/lib/graph/partition.cc [hunk 18]
```diff
@@ -476,24 +478,45 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
+        auto original_type_size = out_edge_types.size();
+        out_edge_types.resize(original_type_size + last - start, m_edge_types[i]);
+        out_edge_weights.reserve(out_edge_weights.size() + last - start);
+        for (size_t index = start; index < last; ++index)
+        {
+            out_edge_weights.emplace_back(index > start ? m_edge_weights[index] - m_edge_weights[index - 1]
+                                                        : m_edge_weights[start]);
+        }
+
+        return last - start;
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:514]: remove

**Review by mortid0 — COMMENTED** [review]
add a description to the changelog.md with the new functionality.

File: src/cc/lib/distributed/graph_engine.cc [hunk 19]
```diff
@@ -324,9 +324,41 @@ grpc::Status GraphEngineServiceImpl::GetEdgeStringFeatures(::grpc::ServerContext
...
+        }
+    }
+
+    return grpc::Status::OK;
+}
+
 grpc::Status GraphEngineServiceImpl::GetNeighbors(::grpc::ServerContext *context,
                                                   const snark::GetNeighborsRequest *request,
                                                   snark::GetNeighborsReply *response)
+
```
Comment (mortid0) [inline src/cc/lib/distributed/graph_engine.cc:361]: nit: remove this new line.

File: src/cc/lib/graph/partition.cc [hunk 20]
```diff
@@ -472,28 +474,48 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
+            result += last - start;
+            func(start, last, i);
         }
     }
     return result;
 }
+
+size_t Partition::NeighborCount(uint64_t internal_id, std::span<const Type> edge_types) const
+{
+    auto lambda = [](auto start, auto last, int i) {};
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:492]: Why 2 first arguments have type auto, but last is int? Be consistent.

File: src/python/deepgnn/graph_engine/snark/tests/neighbor_sampler_test.py [hunk 21]
```diff
@@ -214,6 +214,52 @@ def test_full_neighbor_graph_handle_empty_list(multi_partition_graph_data):
...
+def test_neighbor_count_graph_nonmatching_edge_type(multi_partition_graph_data):
+    g = client.MemoryGraph(multi_partition_graph_data, [0])
+    output_node_counts = g.neighbor_counts(
+        nodes=np.array([9], dtype=np.int64), edge_types=100
+    )
+
+    npt.assert_array_equal(output_node_counts, [0])
+
+
+def test_neighbor_count_graph_nonexistent_node(multi_partition_graph_data):
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/tests/neighbor_sampler_test.py:254]: where are tests for python client in a distributed setting?

File: src/cc/lib/distributed/client.cc [hunk 22]
```diff
@@ -624,6 +624,55 @@ void GRPCClient::GetEdgeStringFeature(std::span<const NodeId> edge_src_ids, std:
...
+
+        call->callback = [&responses_left, &replies, &output_neighbor_counts]() {
+            // Skip processing until all responses arrived. All responses are stored in the `replies` variable,
+            // so we can safely return.
+
+            if (responses_left.fetch_sub(1) > 1)
+            {
+                return;
+            }
+            for (size_t curr_node = 0; curr_node < std::size(output_neighbor_counts); ++curr_node)
```
Comment (mortid0) [inline src/cc/lib/distributed/client.cc:657]: nit: swap loops between line 657 and 660, to process vectors in streaming way:
```c++
for (size_t reply_index = 0; reply_index < std::size(replies); ++reply_index)
{
    const auto& reply = replies[reply[index];
    auto out = std::begin(output_neighbor_counts);
    std::transform (std::begin(output_neighbor_counts), std::end(output_neighbor_counts), std::begin(reply.neighbor_counts()), std::begin(output_neighbor_counts), std::plus<int64_t>());
}
```

File: src/cc/lib/graph/partition.cc [hunk 23]
```diff
@@ -472,28 +474,48 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
         }
     }
     return result;
 }
+
+size_t Partition::NeighborCount(uint64_t internal_id, std::span<const Type> edge_types) const
+{
+    auto lambda = [](auto start, auto last, auto i) {};
+
+    return NeighborIndexIterator(internal_id, edge_types, lambda, m_neighbors_index, m_edge_types, m_edge_type_offset);
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:494]: nit: std::move(lambda)

File: src/cc/lib/graph/partition.cc [hunk 24]
```diff
@@ -472,28 +474,48 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
+            result += last - start;
+            func(start, last, i);
         }
     }
     return result;
 }
+
+size_t Partition::NeighborCount(uint64_t internal_id, std::span<const Type> edge_types) const
+{
+    auto lambda = [](auto start, auto last, auto i) {};
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:492]: nit: auto -> const auto

File: src/cc/lib/graph/partition.cc [hunk 25]
```diff
@@ -472,28 +474,48 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
+        out_edge_types.resize(original_type_size + last - start, m_edge_types[i]);
+        out_edge_weights.reserve(out_edge_weights.size() + last - start);
+        for (size_t index = start; index < last; ++index)
+        {
+            out_edge_weights.emplace_back(index > start ? m_edge_weights[index] - m_edge_weights[index - 1]
+                                                        : m_edge_weights[start]);
+        }
+    };
+
+    return NeighborIndexIterator(internal_id, edge_types, lambda, m_neighbors_index, m_edge_types, m_edge_type_offset);
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:516]: std::move(lambda)

File: examples/pytorch/gat/README.md [hunk 26]
```diff
@@ -1,49 +1,49 @@
-# Introduction
```
Comment (mortid0) [inline examples/pytorch/gat/README.md:1]: Looks like there are lots of non-relevant changes. Could you pick only ones related to neighbor count feature?

File: src/cc/lib/graph/partition.cc [hunk 27]
```diff
@@ -384,21 +345,21 @@ void Partition::GetNodeSparseFeature(uint64_t internal_node_id, std::span<const
                             feature_string.c_str(), node_id_string.c_str());
             continue;
         }
+        assert(stored_size > 12); // minimum is 4 bytes to record there is a single index, actual index (8 bytes)
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:348]: why is this needed?

File: CHANGELOG.md [hunk 28]
```diff
@@ -1,35 +1,38 @@
...
-### Added
-- Add BinaryWriter as new entry point for NodeWriter, EdgeWriter and alias writers.
-
-### Removed
-- Meta.json files are no longer needed by the converter. Remove meta path argument from MultiWorkerConverter and Dispatchers.
-
-### Fixed
-- Fill dimensions with 0 for missing features.
+# Changelog
+All notable changes to this project will be documented in this file.
```
Comment (coledie) [inline CHANGELOG.md:2]: Why does every line have a delta? This whole PR is 13k+/-

Comment (adithyabhonsley) [reply CHANGELOG.md:2]: Think it's because the env was loaded in windows, not linux so files were saved in CRLF format vs LF format. Fixed majority of them

File: src/python/deepgnn/graph_engine/snark/tests/neighbor_sampler_test.py [hunk 29]
```diff
@@ -214,6 +214,126 @@ def test_full_neighbor_graph_handle_empty_list(multi_partition_graph_data):
     npt.assert_array_equal(counts, [0])
 
 
+# Neighbor Count Tests
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/tests/neighbor_sampler_test.py:217]: Remove comment or create a test suite for neighbor count tests.

File: src/python/deepgnn/graph_engine/snark/decoders.py [hunk 30]
```diff
@@ -1,220 +1,220 @@
...
-
-            for i in range(TsvDecoder.NEIGHBOR_COLUMN_COUNT - len(neighbor_columns)):
-                neighbor_columns.append("")
-
-            dst_type = int(neighbor_columns[1]) if neighbor_columns[1] else 0
-            dst_weight = float(neighbor_columns[2]) if neighbor_columns[2] else 0.0
-            yield node_id, dst_id, dst_type, dst_weight, self._parse_feature_string(
-                neighbor_columns[3]
-            )
+# Copyright (c) Microsoft Corporation.
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/decoders.py:1]: Looks like you missed to clean line endings in this file, e.g. look at `edge_string_features` method.

File: src/python/deepgnn/graph_engine/snark/client.py [hunk 31]
```diff
@@ -603,6 +614,35 @@ def edge_string_features(
 
         return py_cb.values, dimensions // py_cb.values.itemsize
 
+    def neighbor_counts(
+        self, nodes: np.ndarray, edge_types: Union[List[int], int]
+    ) -> np.ndarray:
+        """Retrieve degree of node with satisfying edge types.
+
+        Args:
+            nodes -- array of nodes to select neighbors
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/client.py:623]: Please, keep the same format in docstrings as in other methods.

File: src/cc/lib/py_graph.cc [hunk 32]
```diff
@@ -621,6 +621,41 @@ int32_t GetNeighborsInternal(PyGraph *py_graph, NodeID *in_node_ids, size_t in_n
...
 
+int32_t NeighborCount(PyGraph *py_graph, NodeID *in_node_ids, size_t in_node_ids_size, Type *in_edge_types,
+                      size_t in_edge_types_size, uint64_t *out_neighbor_counts)
+{
+    if (py_graph->graph == nullptr)
+    {
+        RAW_LOG_ERROR("Internal graph is not initialized");
+        return 1;
+    }
+    std::fill_n(out_neighbor_counts, in_node_ids_size, 0);
```
Comment (mortid0) [inline src/cc/lib/py_graph.cc:632]: Why do you need to perform `std::_fill` twice? It is done first time here and second time in `graph.cc`/`client.cc`.

File: src/cc/lib/graph/partition.cc [hunk 33]
```diff
@@ -483,28 +438,50 @@ size_t Partition::FullNeighbor(uint64_t internal_id, std::span<const Type> edge_
...
+            result += last - start;
+            func(start, last, i);
         }
     }
     return result;
 }
+
+size_t Partition::NeighborCount(uint64_t internal_id, std::span<const Type> edge_types) const
+{
+    const auto lambda = [](const auto start, const auto last, const auto i) {};
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:456]: What is the purpose of const qualifier here?

Comment (mortid0) [inline src/cc/lib/graph/partition.cc:456]: What is the purpose of const qualifier here?

File: src/cc/lib/graph/partition.cc [hunk 34]
```diff
@@ -435,15 +390,13 @@ void Partition::GetNodeStringFeature(uint64_t internal_node_id, std::span<const
...
         }
-
         const auto data_offset = m_node_feature_index[feature_index_offset + feature];
         const auto stored_size = m_node_feature_index[feature_index_offset + feature + 1] - data_offset;
         // Check if the feature is empty
         if (stored_size == 0)
         {
             continue;
         }
-
```
Comment (mortid0) [inline src/cc/lib/graph/partition.cc:446]: What’s going on with extra formatting?

File: src/cc/lib/distributed/server.cc [hunk 35]
```diff
@@ -83,6 +83,12 @@ class EmptyGraphEngine final : public snark::GraphEngine::Service
         return grpc::Status::OK;
     }
 
+    grpc::Status GetNeighborCounts(::grpc::ServerContext *context, const snark::GetNeighborsRequest *request,
+                                   snark::GetNeighborCountsReply *response) override
+    {
+        return grpc::Status::OK;
```
Comment (mortid0) [inline src/cc/lib/distributed/server.cc:89]: I don't think this implementation will work with the `GRPCClient` defined in `client.cc`. That client expects response counts length to be equal to the number of nodes, but you don't return back anything. Please, modify client.cc and add a test that uses both `GraphEngineServiceImpl` and `EmptyGraphEngine` classes that are connected to a single client.

File: src/cc/lib/distributed/client.cc [hunk 36]
```diff
@@ -624,6 +624,65 @@ void GRPCClient::GetEdgeStringFeature(std::span<const NodeId> edge_src_ids, std:
...
+                return;
+            }
+
+            for (size_t reply_index = 0; reply_index < std::size(replies); ++reply_index)
+            {
+                const auto &reply = replies[reply_index];
+                auto output_len = output_neighbor_counts.size();
+                auto reply_len = reply.neighbor_counts().size();
+
+                // Mismatch in lengths of output and reply vectors
```
Comment (mortid0) [inline src/cc/lib/distributed/client.cc:662]: replace lines 662-675 with:
```C++
std::transform(std::begin(reply.neighbor_counts()), std::begin(reply.neighbor_counts())+std::min(output_len, size_t(reply_len)), std::begin(output_neighbor_counts), std::plus<int64_t>());
```

File: WORKSPACE [hunk 37]
```diff
@@ -17,9 +17,9 @@ http_archive(
 
 http_archive(
     name = "rules_python",
-    sha256 = "a30abdfc7126d497a7698c29c46ea9901c6392d6ed315171a6df5ce433aa4502",
-    strip_prefix = "rules_python-0.6.0",
-    url = "https://github.com/bazelbuild/rules_python/archive/0.6.0.tar.gz",
+    sha256 = "5fa3c738d33acca3b97622a13a741129f67ef43f5fdfcec63b29374cc0574c29",
+    strip_prefix = "rules_python-0.9.0",
```
Comment (mortid0) [inline WORKSPACE:21]: This doesn't seem to be needed anymore.

File: src/cc/lib/distributed/client.cc (hunk 36 above)
Comment (mortid0) [reply src/cc/lib/distributed/client.cc:662]: sorry, you need one more iterator for a binary op:
```C++
std::transform(std::begin(reply.neighbor_counts()), std::begin(reply.neighbor_counts())+std::min(output_len, size_t(reply_len)), std::begin(output_neighbor_counts), std::begin(output_neighbor_counts), std::plus<int64_t>());
```

File: src/python/deepgnn/graph_engine/snark/tests/neighbor_sampler_test.py [hunk 38]
```diff
@@ -294,10 +412,10 @@ def test_remote_client_weighted_sampling_from_unsorted_types(
 def test_remote_client_weighted_sampling_with_missing_neighbors(
     multi_partition_graph_data,
 ):
-    s1 = server.Server(multi_partition_graph_data, [0], "localhost:12346")
-    s2 = server.Server(multi_partition_graph_data, [1], "localhost:12356")
+    s1 = server.Server(multi_partition_graph_data, [0], "localhost:12357")
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/snark/tests/neighbor_sampler_test.py:415]: Please, don't use repeating port numbers between tests.

**Review by mortid0 — APPROVED** [review]

**Review by coledie — APPROVED** [review]
